### PR TITLE
HV-1292 Add support for the new constraints added to the spec

### PR DESCRIPTION
**Pull request by gsmet**
* https://hibernate.atlassian.net/browse/HV-1292
 * https://hibernate.atlassian.net/browse/HV-1293

HV-1293 should probably be backported to all supported releases.

### CODE REVIEW

**Comment by gsmet** [comment]
Jenkins, retest this please

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import javax.validation.constraints.Negative;
+
+/**
+ * Check that the number being validated is negative.
+ *
+ * @author Hardy Ferentschik
+ * @author Xavier Sosnovsky
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: Where is Xavier's name coming from?

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	@Override
+	public boolean isValid(Number value, ConstraintValidatorContext constraintValidatorContext) {
+		// null values are valid
+		if ( value == null ) {
+			return true;
+		}
+		// handling of NaN, positive infinity and negative infinity
+		else if ( value instanceof Double ) {
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: I reckon you won't like it, but what do you think of having dedicated validators for the individual type instaed of the huge instanceof switch here. I'd prefer that, as validator resolution is done only once, so we won't pay the price for checking the type for each validator invocation again and again.

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/notempty/NotEmptyValidatorForArray.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	 * @param array the array to validate
+	 * @param constraintValidatorContext context in which the constraint is evaluated
+	 * @return returns {@code true} if the array is not {@code null} and the array is not empty
+	 */
+	@Override
+	public boolean isValid(Object[] array, ConstraintValidatorContext constraintValidatorContext) {
+		if ( array == null ) {
+			return false;
+		}
+		return Array.getLength( array ) > 0;
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/notempty/NotEmptyValidatorForArray.java]: Interesting, what's the rationale for not using array.length?

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java (hunk 1 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: It's based on DecimalMin/Max validators so I let the name of the original authors.

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java (hunk 2 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: I don't have a strong opinion about it. But I think we should keep consistency. I mimicked DecimalMin/Max so if we decide to change it, we should probably do it globally in a later commit.

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/notempty/NotEmptyValidatorForArray.java (hunk 3 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/notempty/NotEmptyValidatorForArray.java]: Consistency, it's based on the Size validators. But I agree we should use array.length instead. Pushed an additional commit to fix it for Size and NotEmpty.

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java (hunk 1 above)
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: Ah, yes. Makes sense.

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java (hunk 2 above)
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: +1 for splitting up both. Can be done as a follow-up, though, if you prefer.

Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: Yes, I prefer getting the release out of the door ASAP. Will create a new JIRA to track this.

Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: +1

Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/NegativeValidatorForNumber.java]: https://hibernate.atlassian.net/browse/HV-1300

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks, @gsmet!

**Pull request closed without merging**